### PR TITLE
[CUMULUS-2138] Add update-granules-metadata task

### PR DESCRIPTION
**Pull request by botanical**
**Summary:** Summary of changes

Addresses [CUMULUS-2138: As a user I would like the move-granules task to update the metadata files when moveStagedFiles is false](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2138)

## Changes
* Created `update-granules-metadata` task
  * Decouples updating CMR metadata and adding CMR file etags from `move-granules` task and puts them into a new `update-granules-metadata tasks`
  * Updates unit tests on said functionality to account for `.cmr.json` files in addition to `.cmr.xml` files.
* Updates `example/cumulus-tf/ingest_and_publish_granule_workflow.asl.json` to incorporate new `UpdateGranulesMetadata` step and call to `HyraxMetadataUpdatesTask`
* Updates README to point to correct HelloWorld spec
* Updates FileUtils to log when there is an error
* Updates `config.yml` to include my internal bucket
* Updates `example/spec/parallel/ingestGranule/IngestGranuleSuccessSpec.js` to account for added step in workflow

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [ ] Adhoc testing
- [x] Integration tests

### CODE REVIEW

**Comment by botanical** [comment]
Tested locally against `spec/parallel/ingestGranule/IngestGranuleSuccessSpec.js` and `spec/parallel/ingestGranule/IngestUMMGSuccessSpec.js` and both integration tests were successful.

File: tasks/update-granules-metadata/package.json
Comment (markdboyd) [inline tasks/update-granules-metadata/package.json:44]: I don't think either of these packages are used in the code for this task that I can tell?

File: tasks/update-granules-metadata/schemas/config.json
Comment (markdboyd) [inline tasks/update-granules-metadata/schemas/config.json:8]: I'm not seeing where config.collection is used in the task, so I think that can be removed from the config schema?

File: tasks/update-granules-metadata/schemas/config.json
Comment (markdboyd) [inline tasks/update-granules-metadata/schemas/config.json:6]: I also don't think config.bucket is used in the task code?

File: example/cumulus-tf/ingest_and_publish_granule_workflow.asl.json
Comment (markdboyd) [inline example/cumulus-tf/ingest_and_publish_granule_workflow.asl.json:248]: Based on the comments below, I think we might not need `bucket` or `collection` task config for this task

File: tasks/update-granules-metadata/tests/test-index.js
Comment (markdboyd) [inline tasks/update-granules-metadata/tests/test-index.js:99]: I think it might be helpful to have another test where the incoming payload already has an etag on the CMR file and verify that the etag changes after running the task. This scenario would not occur for any of current workflows, but it could if people use this task later in a workflow chain. What do you think?

File: tasks/update-granules-metadata/tests/data/payload.json
Comment (markdboyd) [inline tasks/update-granules-metadata/tests/data/payload.json:23]: I think you might not need this for your task, in which case it'd good to simplify the fixture

File: tasks/update-granules-metadata/tests/test-index.js
Comment (markdboyd) [inline tasks/update-granules-metadata/tests/test-index.js:105]: Since your input payload has two granules, each with a CMR file, why not do this test against the CMR file for both granules?

**Review by markdboyd — CHANGES_REQUESTED** [review]
A few requested changes and comments

File: tasks/update-granules-metadata/tests/test-index.js
Comment (botanical) [reply tasks/update-granules-metadata/tests/test-index.js:105]: Good point, I forgot to add the second test!

File: tasks/update-granules-metadata/tests/test-index.js
Comment (botanical) [reply tasks/update-granules-metadata/tests/test-index.js:99]: @markdboyd 
Yes, I think that is useful to test. I think I need more context on what that workflow could look like. Would a workflow chain in which that scenario might occur look like this:

```
      sync_granule_task_arn: module.cumulus.sync_granule_task.task_arn,
      add_missing_file_checksums_task_arn: 
      module.cumulus.add_missing_file_checksums_task.task_arn,
      fake_processing_task_arn: module.cumulus.fake_processing_task.task_arn,
      files_to_granules_task_arn: module.cumulus.files_to_granules_task.task_arn,
      move_granules_task_arn: module.cumulus.move_granules_task.task_arn,
      hyrax_metadata_updates_task_arn: module.cumulus.hyrax_metadata_updates_task.task_arn,
      post_to_cmr_task_arn: module.cumulus.post_to_cmr_task.task_arn,
      update_granules_metadata_task_arn: module.cumulus.update_granules_metadata_task.task_arn
```
where it updates the granules metadata last? Correct me if I am wrong, but etags are updated on any s3 PUT event, right? So, in this case, hyrax metadata update will write to s3 and it will consequently update the etag. Let me know if there is a fault in my understanding/reasoning!

File: tasks/update-granules-metadata/schemas/config.json
Comment (botanical) [reply tasks/update-granules-metadata/schemas/config.json:8]: Yeah, that's true. I'll remove it

File: tasks/update-granules-metadata/tests/test-index.js
Comment (markdboyd) [reply tasks/update-granules-metadata/tests/test-index.js:105]: I would just update this one test to run the same assertion for the CMR file in both granules

File: tasks/update-granules-metadata/tests/test-index.js
Comment (markdboyd) [reply tasks/update-granules-metadata/tests/test-index.js:99]: > etags are updated on any s3 PUT event

Yes, that is correct. 

That workflow you outlined would trigger the behavior - but I don't think it would ever occur in reality. Once `post_to_cmr ` is run, the metadata on S3 has already been posted to CMR , so there's no need to update the metadata filed stored on S3.

Really, I can't actually think of a real life workflow where update_granules_metadata might run and receive etags as part of the input, but keep in mind that users can assemble our tasks into workflows however they want including with their own custom tasks. So my motivation for testing this is because it is **theoretically** possible for the task to receive input with etags on the files and it seems easy enough to test.

File: tasks/update-granules-metadata/tests/test-index.js
Comment (botanical) [reply tasks/update-granules-metadata/tests/test-index.js:105]: Yeah, that's what I ended up doing

File: tasks/update-granules-metadata/schemas/config.json
Comment (botanical) [reply tasks/update-granules-metadata/schemas/config.json:6]: It's not but I was running into a strange error when I removed it.

File: example/cumulus-tf/ingest_and_publish_granule_workflow.asl.json
Comment (botanical) [reply example/cumulus-tf/ingest_and_publish_granule_workflow.asl.json:248]:
```"CumulusMessageAdapterExecutionError: warning setting command to loadAndUpdateRemoteEvent",
    "warning setting command to loadNestedEvent",
    "Unexpected Error <class 'jsonschema.exceptions.ValidationError'>. config schema: 'collection' is a required property",

```
and
```
"errorType": "CumulusMessageAdapterExecutionError",
  "errorMessage": "warning setting command to loadAndUpdateRemoteEvent\nwarning setting command to loadNestedEvent\nUnexpected Error <class 'jsonschema.exceptions.ValidationError'>. config schema: 'collection' is a required property\n\nFailed validating 'required' in schema:\n    {'description': 'Describes the config used by the '\n                    'update-granules-metadata task',\n     'properties': {'bucket': {'description': 'the bucket the has the '\n                                              'private/public key needed '\n                                              'for decrypting cmr password',\n                               'type': 'string'},\n              
```

Comment (markdboyd) [reply example/cumulus-tf/ingest_and_publish_granule_workflow.asl.json:248]: Yeah, so these errors occurred because you removed them from the `task_config`, which sets the `config.collection` value in the task, but you hadn't removed them from the task config schema. So the value of `config.collection` was empty, but the task config schema was still requiring `config.collection` to be present.

So if you remove `bucket` and `collection` from tasks/update-granules-metadata/schemas/config.json and then also from this `task_config`, I would expect those errors to go away.

File: tasks/update-granules-metadata/tests/data/payload.json
Comment (markdboyd) [inline tasks/update-granules-metadata/tests/data/payload.json:71]: I'm not opposed, but what's the rationale for having three granules in the test fixture as opposed to two? Two is the sufficient minimum case to prove that multiple granules get updated.

File: tasks/update-granules-metadata/tests/test-index.js
Comment (markdboyd) [inline tasks/update-granules-metadata/tests/test-index.js:134]: I know that this is how the test is already written, but it can be refactored to now that we're on a later version of `ava`:

```
  const error = await t.throwsAsync(
    () => updateGranulesMetadata(newPayload),
    { message: 'cmrGranuleUrlType is distribution, but no distribution endpoint is configured.' } 
  );
```

File: tasks/update-granules-metadata/tests/test-index.js
Comment (markdboyd) [inline tasks/update-granules-metadata/tests/test-index.js:118]: Maybe this is personal preference, but even though we're working with a fixture and have a predictable granuleId, I'd prefer that this test find the granules to test on by looking for which ones have an etag. It de-couples the test from knowledge of the fixture.

File: tasks/update-granules-metadata/tests/test-index.js
Comment (markdboyd) [inline tasks/update-granules-metadata/tests/test-index.js:120]: I could be misreading this, but it looks like we're comparing here whether a `file` object found in `granule.files` is not equal to a specific Etag string, so this test will always pass even if the etag isn't being updated.

File: tasks/update-granules-metadata/tests/data/payload.json
Comment (markdboyd) [reply tasks/update-granules-metadata/tests/data/payload.json:71]: I see. I think you added this just for your etag tests? That's fine in that case

Comment (botanical) [reply tasks/update-granules-metadata/tests/data/payload.json:71]: Yeah, I added the third granule for my etag test

File: tasks/update-granules-metadata/tests/test-index.js
Comment (markdboyd) [reply tasks/update-granules-metadata/tests/test-index.js:120]: I think the logic that I would prefer to see for this test is: iterate over granules in input (`newPayload`) and if those granules have files with `etags`, then find the corresponding granule and file in `output` and compare whether those two etags are different

File: example/cumulus-tf/ingest_and_publish_granule_workflow.asl.json
Comment (botanical) [reply example/cumulus-tf/ingest_and_publish_granule_workflow.asl.json]: I removed them from both `task_config` in this file and `schemas/config.json` and still ran into the error, but I realized it's because I didn't remove `bucket` and `collections` from the `required` list in `config.json`... oops 😬

File: tasks/update-granules-metadata/tests/test-index.js
Comment (markdboyd) [inline tasks/update-granules-metadata/tests/test-index.js:122]: Let's use the `t.not` assertion here: https://github.com/avajs/ava/blob/master/docs/03-assertions.md#notvalue-expected-message

File: tasks/update-granules-metadata/tests/test-index.js
Comment (markdboyd) [inline tasks/update-granules-metadata/tests/test-index.js:116]: So the test is still depending on knowing the granuleId from the fixture. I think it could be refactored like this to avoid that dependency:

```
const inputGranulesWithEtags = newPayload.input.granules.reduce(
  (acc, granule) => {
    const cmrFile = granule.files.find(isCMRFile);
    if (!cmrFile) return acc;
    acc.push([granule.granuleId, cmrFile]);
  },
  []
);

inputGranulesWithEtags.forEach(([inputGranuleId, inputCmrFile]) => {
  const outputGranule = output.granules.find((outputGranule) => outputGranule.granuleId === inputGranuleId);
  if (outputGranule) {
    const outputCmrFile = outputGranule.files.find(isCMRFile);
    t.not(inputCmrFile.etag, outputCmrFile.etag)
  }
})
```

What do you think? Does that seem preferable or like unnecessary optimization? Also I didn't test the above code so it could need some tweaking

File: example/cumulus-tf/ingest_and_publish_granule_workflow.asl.json
Comment (markdboyd) [reply example/cumulus-tf/ingest_and_publish_granule_workflow.asl.json]: Ah, makes sense. Glad you figured it out!

File: tasks/update-granules-metadata/tests/test-index.js
Comment (botanical) [reply tasks/update-granules-metadata/tests/test-index.js]: Hm, I did it a little differently. https://github.com/nasa/cumulus/pull/1969/files#diff-41cb5ec44ef17d2ccd2ba9142b73b29aee1d0c497c17766f8877a8d58648aa57R112-R125

Let me know what you think!

File: tasks/update-granules-metadata/tests/test-index.js
Comment (markdboyd) [inline tasks/update-granules-metadata/tests/test-index.js:116]: My only issue is that we're assuming that the first file is the one that has the etag, which is true for the fixture but could change. I think adding the `.filter(isCMRFile)` like you have on line 121 would solve that nicely 💯

Comment (markdboyd) [reply tasks/update-granules-metadata/tests/test-index.js:116]: Also this solution is much more readable than mine, so 💯

Comment (botanical) [reply tasks/update-granules-metadata/tests/test-index.js:116]: Good point! I'll update that

**Review by markdboyd — APPROVED** [review]
This is approved from my perspective, but want to get another team member's input before merging

**Review by nemreid — CHANGES_REQUESTED** [review]
This is a good PR! 💯 I only have some nitpicks, and most importantly a request for more and detailed changelog entries.

File: CHANGELOG.md
Comment (nemreid) [reply CHANGELOG.md:26]: It occurs to me that the files are what's being updated, with URLs, and there is no pushing of CMR metadata to CMR as in PostToCmr, so something like `update-granules-cmr-metadata-file-urls` or `-links`  would be even more specific.

Comment (nemreid) [reply CHANGELOG.md:26]: Naming is hard. 😅

Comment (botanical) [reply CHANGELOG.md:26]: I definitely agree with everything you said. I will consider what we can name this task. I will also add a note under `BREAKING CHANGES` as you have pointed out. Thanks for your feedback!

File: tasks/update-granules-metadata/tests/data/ummg-meta.json
Comment (botanical) [reply tasks/update-granules-metadata/tests/data/ummg-meta.json:61]: This has a newline, doesn't it? 😮

File: CHANGELOG.md
Comment (markdboyd) [reply CHANGELOG.md:26]: We should also look in the docs and see if we need to update any docs where `MoveGranules`/`move-granules` is mentioned

Comment (botanical) [reply CHANGELOG.md]: I've gone with `update-granules-cmr-metadata-file-links` though I do need to acknowledge it is a pretty long name. I find that it most accurately describes what the task does, so I am torn. I think, in general, it's better to name it exactly to describe what it does so I've opted for a long task name despite my hesitations.

File: CHANGELOG.md
Comment (Jkovarik) [inline CHANGELOG.md:35]: Awesome!

File: packages/api/lib/FileUtils.js
Comment (Jkovarik) [inline packages/api/lib/FileUtils.js:75]: Was this intentionally left in?   If so, should we use the logger package, and does it need to be JSON.stringified to be useful given how nested Core file objects are?

File: tasks/update-granules-cmr-metadata-file-links/package.json
Comment (Jkovarik) [inline tasks/update-granules-cmr-metadata-file-links/package.json:15]: I'm not sure we want to publish new tasks, given they're being pulled into TF via our release artifacts.   @nemreid  @markdboyd ?

File: tf-modules/ingest/update-granules-cmr-metadata-file-links-task.tf
Comment (Jkovarik) [inline tf-modules/ingest/update-granules-cmr-metadata-file-links-task.tf:9]: We should probably consider tuning/eval of the timeout here, depending on what load/timing projections looks like for a reasonable amount of granules.   

 Definitely shouldn't block the PR, but move granules had the same setup and was S3 timing bound, system performance/etc will benefit from reducing the timeout for failure cases if we're well above what users would reasonably want.

**Review by Jkovarik — COMMENTED** [review]
This looks great - as it's already approved, I'll defer to the original reviewers, just wanted to note I left a comment for @nemreid  @markdboyd  to review.

File: tasks/update-granules-cmr-metadata-file-links/package.json
Comment (markdboyd) [reply tasks/update-granules-cmr-metadata-file-links/package.json:15]: I think I would defer to publishing until we make a judgment call / change about all the tasks as a whole

File: packages/api/lib/FileUtils.js
Comment (markdboyd) [reply packages/api/lib/FileUtils.js:75]: When we were debugging failing integration tests for Jennifer, it was impossible to see what was really failing with logging some sort of error here. But 👍 to using the logger package

File: tf-modules/ingest/update-granules-cmr-metadata-file-links-task.tf
Comment (botanical) [reply tf-modules/ingest/update-granules-cmr-metadata-file-links-task.tf:9]: Good point! If I understand what you mean by `S3 timing bound`, this task similarly relies on S3 to retrieve an object. That would lead me to think that the timeout for this task is okay, but I am open to reducing the timeout for failure cases like you mentioned. @Jkovarik

File: packages/api/lib/FileUtils.js
Comment (botanical) [reply packages/api/lib/FileUtils.js]: Using logger package now - thanks for the tip

File: docs/data-cookbooks/browse-generation.md
Comment (nemreid) [inline docs/data-cookbooks/browse-generation.md:537]:
```suggestion
Both downstream tasks `UpdateGranulesCmrMetadataFileLinks`, and `PostToCmr` expect a valid CMR file to be available if you want to export to CMR.
```

Correct me if I'm wrong, but MoveGranules no longer does anything with CMR files with this PR?

File: tasks/update-granules-cmr-metadata-file-links/package.json
Comment (nemreid) [inline tasks/update-granules-cmr-metadata-file-links/package.json:3]: guessing this version number needs updating now?

File: tasks/update-granules-cmr-metadata-file-links/package.json
Comment (nemreid) [inline tasks/update-granules-cmr-metadata-file-links/package.json:41]: Also version number updates. I'd expect bootstrap to break when current with master otherwise.

File: tasks/update-granules-cmr-metadata-file-links/tests/data/ummg-meta.json
Comment (nemreid) [inline tasks/update-granules-cmr-metadata-file-links/tests/data/ummg-meta.json:61]: missing newline

**Review by nemreid — COMMENTED** [review]
Appreciate you addressing my previous comments! I think the version number thing is the only thing really of note here since master has a different major version now.

File: docs/data-cookbooks/browse-generation.md
Comment (botanical) [reply docs/data-cookbooks/browse-generation.md:537]: Hm, I don't think that is entirely true. MoveGranules still deals with CMR files in `updateGranulesMetadata()` https://github.com/nasa/cumulus/blob/master/tasks/move-granules/index.js#L79

File: tasks/update-granules-cmr-metadata-file-links/tests/data/ummg-meta.json
Comment (botanical) [reply tasks/update-granules-cmr-metadata-file-links/tests/data/ummg-meta.json:61]: Doesn't it already have a newline? Or does it need another? I will just add another but it seems to have one already.

File: tasks/update-granules-cmr-metadata-file-links/package.json
Comment (botanical) [reply tasks/update-granules-cmr-metadata-file-links/package.json]: yeah, good call out - I've updated

File: tasks/update-granules-cmr-metadata-file-links/package.json
Comment (Jkovarik) [reply tasks/update-granules-cmr-metadata-file-links/package.json:15]: I'm 'ok' with that, but let's get it on the architecture meeting for a discussion.

File: tf-modules/ingest/update-granules-cmr-metadata-file-links-task.tf
Comment (Jkovarik) [reply tf-modules/ingest/update-granules-cmr-metadata-file-links-task.tf:9]: Sorry - what I was (poorly) getting at was generally we probably want timeouts to be trimmed as much as reasonably possible with padding so that config errors/bugs/etc that might cause a timeout don't eat invocation charges at scale -- given what the lambda is doing compared with what move granules was doing, I was wondering if in terms of "wildest reasonable expectations" if 5 minutes makes sense.

If you feel this is reasonable given task testing, no objection at all.  

Again, not a PR stopper at all, it's super minor given the scaling/budgets probably involved.

File: docs/data-cookbooks/browse-generation.md
Comment (nemreid) [reply docs/data-cookbooks/browse-generation.md:537]: OK, then my nitpick is the the use of 'Both' for 3 tasks is slightly odd use of language, but not really a blocking issue.

File: tasks/update-granules-cmr-metadata-file-links/tests/data/ummg-meta.json
Comment (nemreid) [reply tasks/update-granules-cmr-metadata-file-links/tests/data/ummg-meta.json:61]: Screen Shot 2020-11-05 at 9 38 24 AM

I'm seeing Github report a missing newline, and I think it might actually be a JSON formatting issue. Try removing the whitespace before the terminating bracket on the last line?

**Review by nemreid — APPROVED** [review]

**Comment by botanical** [comment]
Going to close because commits have already been merged to branch `merge-cumulus-2138-update-granules-cmr-task`

**Pull request closed without merging**
